### PR TITLE
read zref from power-spectrum

### PR DESCRIPTION
**Pull request by londumas**
Reads `zpk` from input power spectrum.
Remove reference to `zref`.
Initial commit for test, in the next commit, will add saving `zpk` to h5py file.
Fix issue https://github.com/igmhub/picca/issues/367

Gives the same result h5py. I will now adds storing this parameter to the file.

### CODE REVIEW

**Comment by ngbusca** [comment]
@londumas can you take the chance to check that the fit parameters are given at zref from the masurement and not at zpk?

**Comment by mblomqvist** [comment]
Apologies for being a stickler about the names ;) but zref and zpk are by definition the same quantity. The measurement is at zeff. We've been using the zref name for 5+ years, but if you prefer zpk in the code I fine with it.

**Comment by ngbusca** [comment]
@mblomqvist I also like zref better than zpk for the redshift of the fiducial P(k)... 

By the way, if you add Om, etc. to the fiducial P(k) fits file, could you actually add all the parameters in the camb .ini file so we can reproduce the P(k) in 42 years if we need to?

**Comment by mblomqvist** [comment]
@ngbusca Yes, I'd add more parameters. There are a few more obscure ones that I'm not entirely sure about though. In any case, CAMB has different versions and as a result the power spectra may not be exactly identical 42 years from now.

**Comment by londumas** [comment]
@ngbusca, what is the use of `_wrap_chi2`? It doesn't seem to be used.

**Comment by londumas** [comment]
@ngbusca, just changed the name from `zpk` to `zref`.
Do you want me to do anything else?

**Comment by ngbusca** [comment]
@londumas with the current setup, the parameters (bias, etc) are given at `zref`. At some point we should change that so they are given at the redshift of the measurement... what do you think?

**Comment by londumas** [comment]
@ngbusca, I totally agree. But then I have to add a common parameter for all data-set. Not one per data set. We could call it `zeff`? What do you think? I can take care of this.

**Comment by londumas** [comment]
Here is the relation between `z` (redshift of a bin), `zeff` redshift where to give the parameters and `zref` the redshift of the input power spectrum: 
image
image

```
\xi_{biased}(z) = b_1(z_{eff})b_2(z_{eff})
\left(\frac{1+z}{1+z_{eff}}\right)^{\alpha_{1}}
\left(\frac{1+z}{1+z_{eff}}\right)^{\alpha_{2}}
\left(\frac{G(z)}{G(z_{ref})}\right)^{2}
\xi_{Pk}(z_{ref})
```

**Comment by londumas** [comment]
Just as a test, here is the value of `log10(|bias_LYA|)` versus `redshift` for the cross-correlation assuming no evolution of bias_QSO and a `2.9` evolution of `bia_LYA`. Everything is perfect.

capture d ecran de 2018-03-19 13-28-44

**Comment by londumas** [comment]
@ngbusca, ready to be merged.

**Comment by ngbusca** [comment]
@londumas I don't like so much that one has to pass zref to the model functions, because it's a fixed number that doesn't change... but I don't see how else it can be done without more complication. But I really don't like passing zeff. Can you go back to using partial to define the z_evol functions with a fixed zeff? Then you only pass zref to the models.

**Comment by londumas** [comment]
@ngbusca, I can. The issue is that I then need to pass the `zeff` parameter to the `data.data` constructor. Does such a line annoys you: `data.data(parse_data(os.path.expandvars(d)),zeff)` ?
It isn't very beautiful. I can try to pass it trough an other way.

**Comment by londumas** [comment]
@ngbusca, I have another way. I can create a method that acts on the `zevol` and `growth_function`.
Let me do ca commit and you tell me what you think.

**Comment by ngbusca** [comment]
@londumas I think it makes sense to add zeff as an attribute of data. You will need for the output file as well, won't you?

**Comment by londumas** [comment]
@ngbusca, Here is another way to do it.
What do you think of it?

**Comment by ngbusca** [comment]
The thing is that you still have to keep track of zeff in many many places while if you partial it in the beginning it's just only once. This is not minor, since zeff and zref are very similar names, so it's very mistake prone, and since it would cause only minor differences it would be very hard to catch.

**Comment by londumas** [comment]
@ngbusca, did you look at the latest commit? I do `partial` it

**Comment by londumas** [comment]
@ngbusca, sorry, partial them, both zref and zeff

**Comment by mblomqvist** [comment]
I would prefer to save the effective redshift as 'zeff' instead of 'redshift', i.e. g.attrs['zeff'] = self.zeff

**Comment by londumas** [comment]
@mblomqvist, done. @ngbusca, do you want me to change other things?

**Comment by londumas** [comment]
Added fixing issue https://github.com/igmhub/picca/issues/375.
@ngbusca, Could you review the code for setting the parameters `Om` and `OL`?
I am not sure to understand how cached and uncached function works

**Comment by londumas** [comment]
This is one solution, the other is to add the `Om=Om, OL=OL` parameters to `growth_factor_no_de`.
As you prefer @ngbusca.
```
        if dic_init['model']['growth function'] in ['growth_factor_de','cached_growth_factor_de']:
            self.growth_function = partial(getattr(xi, dic_init['model']['growth function']),zref=zref, Om=Om, OL=OL)
        else:
            self.growth_function = partial(getattr(xi, dic_init['model']['growth function']),zref=zref)
```

**Comment by ngbusca** [comment]
@londumas note that `growth_factor_de` is already cached. `cached_growth_factor_de` should never be specified by the user. I prefer the solution with partial...

**Comment by londumas** [comment]
@ngbusca, is that ok now?

File: py/picca/fitter2/parser.py [hunk 1]
```diff
@@ -75,15 +82,17 @@ def parse_data(filename):
...
 
     dic_init['cuts'] = {}
     for item, value in cp.items('cuts'):
         dic_init['cuts'][item] = float(value)
 
     dic_init['model'] = {}
+    dic_init['model']['zeff'] = zeff
+    dic_init['model']['zref'] = fiducial['zref']
+    dic_init['model']['Om']   = fiducial['Om']
+    dic_init['model']['OL']   = fiducial['OL']
```
Comment (ngbusca) [inline py/picca/fitter2/parser.py]: I'm surprised that coadacy doesn't complaint about these extra spaces. Could you please remove the double spaces here?

File: py/picca/fitter2/xi.py [hunk 2]
```diff
@@ -142,21 +142,18 @@ def wrapper(*args, **kwargs):
 def growth_factor_de(*args, **kwargs):
     return cached_growth_factor_de(*args, **kwargs)
 
-def cached_growth_factor_de(z, zref = None, **kwargs):
+def cached_growth_factor_de(z, zref = None, Om = None, OL = None, **kwargs):
```
Comment (ngbusca) [inline py/picca/fitter2/xi.py]: Same here. There shouldn't be spaces around the `=` in function parameters

File: py/picca/fitter2/parser.py (hunk 1 above)
Comment (londumas) [reply py/picca/fitter2/parser.py]: done

File: py/picca/fitter2/xi.py (hunk 2 above)
Comment (londumas) [reply py/picca/fitter2/xi.py]: done

**Comment by londumas** [comment]
@ngbusca, just applied your request.

**Comment by ngbusca** [comment]
@londumas there are conflicts in this branch, can you take a look?

**Comment by londumas** [comment]
@ngbusca, corrected.

**Pull request merged**